Add scons/3.1.2 recipe

madebr opened the pull request:
Specify library name and version:  **scons/3.1.2**

This will probably fail on Visual Studio because of an error while linking the test_package:
```
scons: Entering directory `C:\Users\maarten\source\repos\conan-center-index\recipes\scons\all\test_package'
scons: Reading SConscript files ...
conan-compiler tool is msvc
CC is: cl
scons: done reading SConscript files.
scons: Building targets ...
scons: building associated VariantDir targets: build\882726301a296ee1f7f1d98a3db2003150ea70a8
cl /Fobuild\882726301a296ee1f7f1d98a3db2003150ea70a8\test_package.obj /c test_package.c /nologo
test_package.c
link /nologo /OUT:build\882726301a296ee1f7f1d98a3db2003150ea70a8\test_package.exe /LIBPATH:C:\Users\maarten\.conan\data\scons\3.1.2\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\lib build\882726301a296ee1f7f1d98a3db2003150ea70a8\test_package.obj
LINK : fatal error LNK1104: cannot open file 'C:\WINDOWS\lnk{ECD57055-886D-4616-BDF6-EE1EDE9390CA}.tmp'
scons: *** [build\882726301a296ee1f7f1d98a3db2003150ea70a8\test_package.exe] Error 1104
scons: building terminated because of errors.
ERROR: scons/3.1.2 (test package): Error in build() method, line 29
        self.run("scons -C \"{}\" {} --conan-compiler={}".format(self.source_folder, debug_opt, self._script_tool[str(self.settings.compiler)]))
        ConanException: Error 2 while executing scons -C "C:\Users\maarten\source\repos\conan-center-index\recipes\scons\all\test_package"  --conan-compiler=msvc
```

This is the same error as this bug report: https://scons-users.scons.narkive.com/FRzlVB7H/msvc-error
I think this is a bug in either scons or msvc.

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

madebr commented:
~Works on my pc~ ==> Works on CI :tm: :sweat_smile:

SSE4 commented on recipes/scons/all/test_package/SConscript:
```diff
@@ -0,0 +1,40 @@
+env_tools = None
+
+env = Environment(tools=env_tools,
+                  #CC=os.environ.get("CC", None),
```
pls remove dead code

SSE4 commented on recipes/scons/all/conanfile.py:
```diff
@@ -0,0 +1,69 @@
+                        os.path.join(self.package_folder, "lib"))
+
+        # Check for compiled python sources
+        for root, _, files in os.walk(self.package_folder):
```
why this is needed?

uilianries replied:
I think it should be moved to the test package.

madebr replied:
Yeah, good idea.

madebr replied:
I remember why I put it in `package`:
python automatically byte compiles modules as it uses them.
If I do this in the `package` method of the recipe itself, no such objects will be in the folder.
But this test is done in the test_package, a previous run of scons will cause this test to fail.
Running `conan test test_package scons/3.1.2@ -tbf /tmp/scons_tp` twice will not work then.

SSE4 replied:
maybe it's better to remove that files, instead of failing the build?
also, I suppose package method support filtering (** excludes**)

madebr replied:
`self.copy` has an *excludes* option.
The recipe is not using `self.copy` but install of `setup.py`.
But I'll change it to: remove the `.py[cdow]` and print a warning.

SSE4 replied:
yeah, much better

danimtb replied:
Is there something left to do with this?

madebr replied:
I don't think so. @SSE4 seems to be happy 🤣 .
I'm resolving this 😄 .

SSE4 commented on recipes/scons/all/conanfile.py:
```diff
@@ -0,0 +1,69 @@
+                        os.path.join(self.package_folder, "bin"))
+            shutil.move(os.path.join(self.package_folder, "Lib"),
+                        os.path.join(self.package_folder, "lib2"))
+            shutil.move(os.path.join(self.package_folder, "lib2"),
```
so, you rename "Lib" -> "lib2", then "lib2" -> "lib"? is it just to have lower-case directory after the rename?
I believe you may keep the initial layout project has, just add `cpp_info.libdirs` / `cppinfo.bindirs`?

madebr replied:
Yes, I could do that. But isn't it nicer to have a lib folder, like all recipes have?

madebr replied:
I set the libdirs to an empty directory because I do not want these dirs added to the include path of dependencies.

SSE4 replied:
okay, looks good, just add some comment(s) so people know what's going on there

madebr replied:
Added some documentation.

SSE4 commented on recipes/scons/all/conanfile.py:
```diff
@@ -0,0 +1,69 @@
+        self.copy("LICENSE*", src=self._source_subfolder, dst="licenses")
+
+        # Mislead CI and create an empty header in the includue directory
+        include_dir = os.path.join(self.package_folder, "include")
```
maybe just put an exception to the hook's whitelist?

madebr replied:
@uilianries Is this workaround ok? Or does it need to be added to the hook whitelist?

SSE4 replied:
okay, let's leave it for now as is

danimtb commented on recipes/scons/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+        self.run("{} --version".format(scons_path), run_environment=True, output=output)
+        text = output.getvalue()
+        print(text)
+        assert self.requires["scons"].ref.version in text
```
Although I think none of them are documented, I think it will be *safer* get the version from `self.deps_cpp_info["scons"].version`

madebr replied:
I have addressed your concerns. Please review again.

madebr replied:
Ok, https://github.com/conan-io/conan-center-index/pull/814#issuecomment-598255464 happens so I revert it back to the original.

danimtb replied:
This should be addressed since Conan 1.24 https://github.com/conan-io/conan/releases/tag/1.24.0

We can merge this and apply the fix later, just to keep in mind

danimtb commented on recipes/scons/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+        output = StringIO()
+        self.run("{} --version".format(scons_path), run_environment=True, output=output)
+        text = output.getvalue()
+        print(text)
```
remove print or use `self.output.info(".....")`

madebr replied:
```suggestion
```

madebr commented on recipes/scons/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,27 @@
+        self.run("{} --version".format(scons_path), run_environment=True, output=output)
+        text = output.getvalue()
+        print(text)
+        assert self.requires["scons"].ref.version in text
```
```suggestion
        assert self.deps_cpp_info["scons"].version in text
```

madebr commented on recipes/scons/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,26 @@
+        output = StringIO()
+        self.run("{} --version".format(scons_path), run_environment=True, output=output)
+        text = output.getvalue()
+        assert self.deps_cpp_info["scons"].version in text
```
```suggestion
        assert self.requires["scons"].ref.version in text
```

madebr commented:
> Does it make sense? would it be a valid approach without using private items?

The only reason I'm using `self.requires["scons"].ref.version` in `test_package` is to be sure the version of the executed scons is the same as the packaged one.
I don't think it makes sense to add this to `user_info` only for its `test_package`.

I think it is weird that there is currently no *official* way to get the name and version of the package a test_package is supposed to test.

jgsogo commented:
> I think it is weird that there is currently no _official_ way to get the name and version of the package a test_package is supposed to test.

I totally agree with you. We need that feature, maybe it is time to propose it. It is easy to implement (probably just a matter of documenting, so there'd be no need to wait until the next release)

danimtb commented:
Waiting for the feature in order to update and merge this PR: conan-io/conan#6723

Will be released in Conan 1.24

SSE4 commented:
@madebr @danimtb I think 1.24 was released, may we resume work here?

a4z commented:
since I would already need that packages, if I can help somehow, please let me know

madebr commented:
@danimtb 
Please unblock + review + test + merge + ... :smile:

uilianries commented on recipes/scons/all/test_package/.gitignore:
```diff
@@ -0,0 +1 @@
+*.dblite
```
Remove the gitignore file

madebr replied:
scons creates a `.sconsign.dblite`  file when building.
The `.gitignore` is there to avoid `git` picking it up.

uilianries replied:
so change the root gitignore in CCI repo

uilianries replied:
@madebr in SCons manual, sconsign.dblite can be changed, so you can move it to test_package/<hash>/build folder: https://scons.org/doc/3.1.2/HTML/scons-man.html

madebr replied:
In the mean time, `*.dblite` has been added to  to the global `.gitignore` in ba2b7bf8e.
But I fixed the location of this signature file thanks to your tip.

uilianries commented:
@madebr , You could open a separated PR for the gitignore, otherwise, we CCI bot won't run this build.

a4z commented:
I have used this package to build real life scons projects on windows linux and mac, works excellent! 
when will this go into master? I think the git ignore stuff was  solved, can you please recheck, @jgsogo ?
Sorry for the stress, but I would like to make this useable to our team and then is is not possible to point it to my fork and mymaster

jgsogo commented:
Hi, this package has a fundamental problem: it is a python application, and I'm afraid of the implications.

When you create this package it is running: `self.run("{} setup.py build -j{}".format(sys.executable, tools.cpu_count()))`, it can install any Python requirement into your environment and break Conan itself (best thing, it would just fail to create). Now it is working because `scons` doesn't have dependencies, but that might not always be the same.

At runtime, the same problem: if this package has dependencies, they won't be installed in the machine running the application and it will fail.... even worse if you run Conan using an installer and you don't have Python in the system.

What are your thoughts about these couple of things? Are they important? Will they hit many users? Can we do something to prevent them?

Thanks!

madebr commented:
Add `{python} setup.py  --requires` to the build step and fail if the result is not empty.
Would this be fine?

Thinking about it, does this recipe work if the windows `conan.exe` executable wrapper is used?
Adding a `cpython` recipe (in the future) would fix this.

uilianries commented on recipes/scons/all/conanfile.py:
```diff
@@ -19,12 +21,23 @@ class SConsConan(ConanFile):
+    def configure(self):
+        # Detect a frozen conan PyInstaller
+        if getattr(sys, "frozen", False):
+            raise ConanException("This recipe requires a python interpreter and does not support a frozen conan installation. This can be fixed with a cpython recipe.")
```
```suggestion
            raise ConanInvalidConfiguration("This recipe requires a python interpreter and does not support a frozen conan installation. This can be fixed with a cpython recipe.")
```

ConanException breaks the build

madebr commented on recipes/scons/all/conanfile.py:
```diff
@@ -0,0 +1,88 @@
+from conans import ConanFile, tools
+from conans.errors import ConanException
```
```suggestion
from conans.errors import ConanInvalidConfiguration
```

SSE4 commented on recipes/scons/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,28 @@
+        if self.deps_cpp_info["scons"].version not in text:
+            raise ConanException("scons --version does not return correct version")
+
+        self.output.info("TMP={}".format(os.environ.get("TMP")))
```
I am not sure why do we need to print TMP here, but okay...

madebr replied:
I have an issue running this on my Windows machine. See https://github.com/conan-io/conan-center-index/pull/814#issuecomment-583888754
At one point, there was some problem with temporary folders.
This is a leftover from debugging.
I'll remove it.

madebr commented on recipes/scons/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,28 @@
+        self.run("{} --version".format(scons_path), run_environment=True, output=output)
+        text = output.getvalue()
+        if self.deps_cpp_info["scons"].version not in text:
+            raise ConanException("scons --version does not return correct version")
+
+        self.output.info("TMP={}".format(os.environ.get("TMP")))
```
```suggestion
            raise ConanException("scons --version does not return correct version")
```

madebr replied:
```suggestion
            raise ConanException("scons --version does not return correct version")
```

fulara commented:
I have comment about this:
```
    def configure(self):
        # Detect a frozen conan PyInstaller
        if getattr(sys, "frozen", False):
            raise ConanInvalidConfiguration("This recipe requires a python interpreter and does not support a frozen conan installation. This can be fixed with a cpython recipe.")
```
I am not sure what this exactly means... but it seems because i am bundling conan with pyinsstaller this recipe is not buildable for me?
Is that expected?
Comment above says:
```
Thinking about it, does this recipe work if the windows conan.exe executable wrapper is used?
Adding a cpython recipe (in the future) would fix this.
``` 
so should that check be windows only?

I have removed the throw, changes {} sys.executable to `python` and this compiles now.

Well with minor remark that actually test_package does not compile .. but i'll just disable that one.

madebr commented:
> I am not sure what this exactly means... but it seems because i am bundling conan with pyinsstaller this recipe is not buildable for me?
> Is that expected?

See https://pyinstaller.readthedocs.io/en/v3.3.1/runtime-information.html#run-time-information
the scons recipe is python-code only so needs a standalone `python` interpreter.

> so should that check be windows only?

pyinstaller can also create standalone executables for linux.
You can never be 100% certain that a python interpreter is present, even on linux.
But a check such as `tools.which("python") is None` would also be fine.

> I have removed the throw, changed {} sys.executable to `python` and this compiles now.

Yes, because you have a python interpreter installed.

> Well with minor remark that actually test_package does not compile .. but i'll just disable that one.

I've changed the `getattr(sys, "frozen", False)` check.

fulara commented:
@madebr 
I also needed to change your sys.executable to just plain invocations of `python`.
this is because when conan is bundled with the pyinstaller the sys.executable is `conan`.
So the build fails.

Could this scenario be handled?

fulara commented on recipes/scons/all/conanfile.py:
```diff
@@ -34,7 +34,7 @@ def source(self):
         with tools.chdir(self._source_subfolder):
             output = io.StringIO()
-            self.run("{} setup.py --requires".format(sys.executable), output=output)
+            self.run("python setup.py --requires", output=output)
```
there is few more places that do this 
(cant comment there from view i have open, but line 42 and 53 )

madebr replied:
Thanks for checking and running the recipe.
Is this also used on your *ancient* build system?

fulara replied:
haha i actually lol'd.
i am still battling that subversiong thing, yes.
and thanks to your other comment i noticed my subversion build does not handle http -_-.
and serf library which is used by subversion (presumably, didnt get to use it yet ) ofc uses something called scons I just heard legends about!!

Btw, If i was doing what i am trying to do ~3months ago i'll be doomed, i should send you flowers.

fulara commented on recipes/scons/all/conanfile.py:
```diff
@@ -0,0 +1,93 @@
+
+    @property
+    def _python_executable(self):
+        return "python"
```
should this be 'python3' ? scons requires at least python 3.5

fulara replied:
nvm, scons requires python to be named 'python' i think.

madebr replied:
This is where a python recipe will come in handy.
I have one for 3.7 but it's not working for WIndows right now.
And I should update it for 3.8.

uilianries replied:
Is there any limitation using sys.executable here?

madebr replied:
No, the current recipe does not use `sys.executable` at all.
It assumes `python` is present.

Maybe it requires a `cpython` (build) requirement?

SSE4 replied:
this doesn't align with how existing recipes are using python (such as boost, pybind, meson, etc.). I believe most of users already have python and they expect the same interpreter to be used. I think (it might be completely wrong) usage of .exe installer for conan is very low, and most people install it via pip, so python is already there.

madebr replied:
I side with @SSE4 .
People who use the windows conan exe installer, should add cpython to their profile as a build requirement.
Currently, the same goes for people not having cmake installed.

madebr replied:
What I don't know, is whether python is always present.
Now python2 is EOF, python3 is in use.
Do newer linux distribution always have `python` present? Or should it use `python3`?

uilianries replied:
> People who use the windows conan exe installer, should add cpython to their profile as a build requirement.
Currently, the same goes for people not having cmake installed.

I agree.

> Do newer linux distribution always have python present?

Yes, Python3 as Python 2 is EOL. I would prefer call "python" in this case, because is calls the default one. In my case, I have both installed, but I use `alias python=python3`, otherwise, my system uses python2 as default.

madebr replied:
To add another data point, the meson recipe also implicitly requires a python interpreter.

SSE4 replied:
this is totally okay, there are plenty of recipes (include meson) which rely on python. it's definitely correct assumption, as conan itself requires python, and Linux/OSX usually have python out of the box anyway. 
we have similar assumptions about other tools - e.g. we usually believe that shell (bash), make, perl, autotools are available on *nix systems. we also assume cmake is available everywhere.
so for me it's not a big deal, otherwise we might need to change almost everything to the paranoid mode. let's be practical here.

madebr commented:
Should this recipe (optionally) depend on #1510 for a build requirement?
Should we expect users to have a python interpreter installed and available?
Users of the standalone windows conan.exe can be expected to add cpython to the build_requirements section of its profile.

madebr commented:
@uilianries @danimtb @SSE4 @jgsogo 
Is something holding this back?
I think I addressed the issue about potentially downloading external dependencies.
